js/modules/loki_message_api.js

```diff
@@ -78,19 +79,32 @@ class LokiMessageAPI {
   async sendMessage(pubKey, data, messageTimeStamp, ttl, options = {}) {
     const {
       isPing = false,
+      isPublic = false,
       numConnections = DEFAULT_CONNECTIONS,
-      publicEndpoint = null,
+      channelSettings = null,
     } = options;
     // Data required to identify a message in a conversation
     const messageEventData = {
       pubKey,
       timestamp: messageTimeStamp,
     };
 
-    // FIXME: should have public/sending(ish hint) in the option to make
-    // this more obvious...
-    if (publicEndpoint) {
+    if (isPublic) {
       // could we emit back to LokiPublicChannelAPI somehow?
+      const { server, channelId, conversationId } = channelSettings;
+      const serverAPI = lokiPublicChatAPI.findOrCreateServer(server);
+      const token = await serverAPI.getServerToken();
+      if (!token) {
+        throw new window.textsecure.PublicChatError(
+          `Failed to retrieve valid token for ${conversationId}`
+        );
+      }
+      const channelAPI = serverAPI.findOrCreateChannel(
+        channelId,
+        conversationId
+      );
```

Comment:
Hm we could probably go back to attaching the public endpoint and token to the message earlier in the sending pipeline instead of the channel settings used to get the server api...
I think the reason I did it this was was because if we want to be able to show an error about the message failing to send if the token is invalid, it has to happen at this point after the message has been put in the sending cache and all that stuff
So it just felt strange to get the token at the start, know that it was undefined, and still pass it along till this point before throwing an error, might still be better than this tho

Reply:
Well we need the message to get to a certain point in the sending pipeline before it can be reflected to the user as a sending error, if we fail too early (which I was doing initially with the token) the UI won't update at all
I think I can make this work a bit cleaner tho, I'll push a commit in a min